src/java.base/share/classes/java/io/PrintStream.java

```diff
@@ -818,6 +828,9 @@ private void implWrite(String s) throws IOException {
     // using println, but since subclasses could exist which depend on
     // observing a call to print followed by newLine we only use this if
     // getClass() == PrintStream.class to avoid compatibility issues.
+    /*[IF CRIU_SUPPORT]*/
+    @NotCheckpointSafe
```

Comment:
Why should this not apply to `writeln(char[])` as well?
What about other methods that invoke `lock.lock()`, e.g. `close()` and `flush()`?
Might a better place for this be on `InternalLock.lock()`?

Reply:
The `@NotCheckpointSafe` annotation is to delay taking a checkpoint (for a certain amount of attempts) while a non checkpoint thread owns the global resource.
While taking a checkpoint, single threaded mode is enabled, i.e., only current checkpoint thread is active, other Java threads are paused.
* https://github.com/eclipse-openj9/openj9/issues/18471

In this issue, non-checkpoint thread took the `lock` at https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/2278b32d9391765c7736be31ff94dac09fd56cbb/src/java.base/share/classes/java/io/PrintStream.java#L839, and was paused at single threaded mode.
When the checkpoint thread attempted to acquire the `lock` at https://github.com/ibmruntimes/openj9-openjdk-jdk21/blob/2278b32d9391765c7736be31ff94dac09fd56cbb/src/java.base/share/classes/java/io/PrintStream.java#L837, parked there indefinitely and a `JVMCheckpointException` was thrown.

Reply:
That isn't very reassuring: waiting until users have problems before fixing the code. Shouldn't we be more proactive?

Reply:
This issue is specific to the application, which is tailored to JVM CRIU currently. 
It is intermittent, and only occurs at CRIU checkpoint or restore single-threaded mode stage. It won't happen whenever the application starts running.
JVM throws `JVMCRIUException` to indicate the code causing the issue.